File: 48-rotate-image/rotate-image.py
# Brute Force Code & Optimal Code
class Solution:
    def rotate(self, matrix: List[List[int]]):
        # size of the square matrix
        n = len(matrix)
        # compute the transpose of the matrix swap matrix[i][j] with matrix[j][i] for all element on and above the main diagonal
        for i in range(n):
            for j in range(i, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        # reverse every row
        for i in range(n):
            matrix[i].reverse()

# Time Complexity : O(N^2)
# Space Complexity : O(N)

File: 48-rotate-image/rotate-image.cpp
// Brute Force Code & Opitmal Code
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // size of the square matrix
        int n = matrix.size();
        // compute the transpose of the matrix swap matrix[i][j] with matrix[j][i] for all elements on and above the main diagonal
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }
        // reverse every row
        for (int i = 0; i < n; i++) {
            reverse(matrix[i].begin(), matrix[i].end());
        }
    }
};

// Time Complexity : O(N^2)
// Space Complexity : O(N)